File: src/main/java/com/osscheduler/Scheduler.java
package com.osscheduler;

import java.util.*;

public class Scheduler {

    // ─── FCFS ────────────────────────────────────────────────────────────────
    public static ScheduleResult fcfs(List<Process> input) {
        List<Process> processes = deepCopy(input);
        processes.sort(Comparator.comparingInt(Process::getArrivalTime));

        List<int[]> gantt = new ArrayList<>();
        int time = 0;

        for (Process p : processes) {
            if (time < p.getArrivalTime())
                time = p.getArrivalTime();
            gantt.add(new int[] { p.getPid(), time, time + p.getBurstTime() });
            time += p.getBurstTime();
            p.setCompletionTime(time);
            p.computeMetrics();
        }
        return new ScheduleResult("FCFS", processes, gantt);
    }

    // ─── SJF (Non-Preemptive) ────────────────────────────────────────────────
    public static ScheduleResult sjf(List<Process> input) {
        List<Process> processes = deepCopy(input);
        List<Process> remaining = new ArrayList<>(processes);
        List<int[]> gantt = new ArrayList<>();
        int time = 0;
        int completed = 0;

        while (completed < processes.size()) {
            List<Process> ready = new ArrayList<>();
            for (Process p : remaining) {
                if (p.getArrivalTime() <= time)
                    ready.add(p);
            }

            if (ready.isEmpty()) {
                time++;
                continue;
            }

            ready.sort(Comparator.comparingInt(Process::getBurstTime));
            Process p = ready.get(0);
            remaining.remove(p);

            gantt.add(new int[] { p.getPid(), time, time + p.getBurstTime() });
            time += p.getBurstTime();
            p.setCompletionTime(time);
            p.computeMetrics();
            completed++;
        }
        return new ScheduleResult("SJF", processes, gantt);
    }

    // ─── Priority (Non-Preemptive, lower number = higher priority) ──────────
    public static ScheduleResult priority(List<Process> input) {
        List<Process> processes = deepCopy(input);
        List<Process> remaining = new ArrayList<>(processes);
        List<int[]> gantt = new ArrayList<>();
        int time = 0;
        int completed = 0;

        while (completed < processes.size()) {
            List<Process> ready = new ArrayList<>();
            for (Process p : remaining) {
                if (p.getArrivalTime() <= time)
                    ready.add(p);
            }

            if (ready.isEmpty()) {
                time++;
                continue;
            }

            ready.sort(Comparator.comparingInt(Process::getPriority));
            Process p = ready.get(0);
            remaining.remove(p);

            gantt.add(new int[] { p.getPid(), time, time + p.getBurstTime() });
            time += p.getBurstTime();
            p.setCompletionTime(time);
            p.computeMetrics();
            completed++;
        }
        return new ScheduleResult("Priority", processes, gantt);
    }

    // ─── Round Robin ─────────────────────────────────────────────────────────
    public static ScheduleResult roundRobin(List<Process> input, int quantum) {
        List<Process> processes = deepCopy(input);
        processes.sort(Comparator.comparingInt(Process::getArrivalTime));

        Queue<Process> queue = new LinkedList<>();
        List<int[]> gantt = new ArrayList<>();
        int time = 0;
        int idx = 0; // index into sorted processes list

        // add processes that arrive at time 0
        while (idx < processes.size() && processes.get(idx).getArrivalTime() <= time) {
            queue.add(processes.get(idx++));
        }

        while (!queue.isEmpty()) {
            Process p = queue.poll();
            int execTime = Math.min(quantum, p.getRemainingTime());

            gantt.add(new int[] { p.getPid(), time, time + execTime });
            time += execTime;
            p.setRemainingTime(p.getRemainingTime() - execTime);

            // enqueue newly arrived processes
            while (idx < processes.size() && processes.get(idx).getArrivalTime() <= time) {
                queue.add(processes.get(idx++));
            }

            if (p.getRemainingTime() > 0) {
                queue.add(p);
            } else {
                p.setCompletionTime(time);
                p.computeMetrics();
            }
        }
        return new ScheduleResult("Round Robin (Q=" + quantum + ")", processes, gantt);
    }

    // ─── Helper ──────────────────────────────────────────────────────────────
    private static List<Process> deepCopy(List<Process> original) {
        List<Process> copy = new ArrayList<>();
        for (Process p : original) {
            copy.add(new Process(p.getPid(), p.getArrivalTime(), p.getBurstTime(), p.getPriority()));
        }
        return copy;
    }

    // ─── SRTF (Preemptive SJF) ────────────────────────────────────────────────
    public static ScheduleResult srtf(List<Process> input) {
        List<Process> processes = deepCopy(input);
        int n = processes.size();
        int time = 0, completed = 0;
        List<int[]> gantt = new ArrayList<>();
        int lastPid = -1;
        int segStart = 0;

        while (completed < n) {
            Process shortest = null;
            for (Process p : processes) {
                if (p.getArrivalTime() <= time && p.getRemainingTime() > 0) {
                    if (shortest == null || p.getRemainingTime() < shortest.getRemainingTime())
                        shortest = p;
                }
            }

            if (shortest == null) {
                time++;
                continue;
            }

            if (shortest.getPid() != lastPid) {
                if (lastPid != -1)
                    gantt.add(new int[] { lastPid, segStart, time });
                segStart = time;
                lastPid = shortest.getPid();
            }

            shortest.setRemainingTime(shortest.getRemainingTime() - 1);
            time++;

            if (shortest.getRemainingTime() == 0) {
                shortest.setCompletionTime(time);
                shortest.computeMetrics();
                completed++;
                gantt.add(new int[] { shortest.getPid(), segStart, time });
                lastPid = -1;
            }
        }
        return new ScheduleResult("SRTF", processes, gantt);
    }

}

File: src/main/webapp/react-app/src/scheduler.js
// Deep copy processes with computed fields reset
function dc(ps) {
  return ps.map(p => ({ ...p, remainingTime: p.burstTime, completionTime: 0, turnaroundTime: 0, waitingTime: 0 }))
}
function cm(p) {
  p.turnaroundTime = p.completionTime - p.arrivalTime
  p.waitingTime    = p.turnaroundTime - p.burstTime
}
function mk(algorithm, processes, ganttChart) {
  const avgWT  = processes.reduce((s, p) => s + p.waitingTime, 0)  / processes.length
  const avgTAT = processes.reduce((s, p) => s + p.turnaroundTime, 0) / processes.length
  return { algorithm, processes, ganttChart, avgWT, avgTAT }
}

export function fcfs(input) {
  const ps = dc(input).sort((a, b) => a.arrivalTime - b.arrivalTime)
  const g = []; let t = 0
  ps.forEach(p => {
    if (t < p.arrivalTime) t = p.arrivalTime
    g.push({ pid: p.pid, start: t, end: t + p.burstTime })
    t += p.burstTime; p.completionTime = t; cm(p)
  })
  return mk('FCFS', ps, g)
}

export function sjf(input) {
  const ps = dc(input), rem = [...ps], g = []; let t = 0, done = 0
  while (done < ps.length) {
    const ready = rem.filter(p => p.arrivalTime <= t)
    if (!ready.length) { t++; continue }
    ready.sort((a, b) => a.burstTime - b.burstTime)
    const p = ready[0]; rem.splice(rem.indexOf(p), 1)
    g.push({ pid: p.pid, start: t, end: t + p.burstTime })
    t += p.burstTime; p.completionTime = t; cm(p); done++
  }
  return mk('SJF', ps, g)
}

export function prioritySched(input) {
  const ps = dc(input), rem = [...ps], g = []; let t = 0, done = 0
  while (done < ps.length) {
    const ready = rem.filter(p => p.arrivalTime <= t)
    if (!ready.length) { t++; continue }
    ready.sort((a, b) => a.priority - b.priority)
    const p = ready[0]; rem.splice(rem.indexOf(p), 1)
    g.push({ pid: p.pid, start: t, end: t + p.burstTime })
    t += p.burstTime; p.completionTime = t; cm(p); done++
  }
  return mk('Priority', ps, g)
}

export function roundRobin(input, quantum) {
  const ps = dc(input).sort((a, b) => a.arrivalTime - b.arrivalTime)
  const queue = [], g = []; let t = 0, idx = 0
  while (idx < ps.length && ps[idx].arrivalTime <= t) queue.push(ps[idx++])
  while (queue.length) {
    const p = queue.shift()
    const ex = Math.min(quantum, p.remainingTime)
    g.push({ pid: p.pid, start: t, end: t + ex })
    t += ex; p.remainingTime -= ex
    while (idx < ps.length && ps[idx].arrivalTime <= t) queue.push(ps[idx++])
    if (p.remainingTime > 0) queue.push(p)
    else { p.completionTime = t; cm(p) }
  }
  return mk(`Round Robin (Q=${quantum})`, ps, g)
}

export function srtf(input) {
  const ps = dc(input)
  const n = ps.length
  let t = 0, completed = 0
  const g = []
  let lastPid = -1, segStart = 0

  while (completed < n) {
    const ready = ps.filter(p => p.arrivalTime <= t && p.remainingTime > 0)
    if (!ready.length) { t++; continue }
    ready.sort((a, b) => a.remainingTime - b.remainingTime)
    const p = ready[0]

    if (p.pid !== lastPid) {
      if (lastPid !== -1) g.push({ pid: lastPid, start: segStart, end: t })
      segStart = t; lastPid = p.pid
    }

    p.remainingTime--; t++
    if (p.remainingTime === 0) {
      p.completionTime = t; cm(p); completed++
      g.push({ pid: p.pid, start: segStart, end: t })
      lastPid = -1
    }
  }
  return mk('SRTF', ps, g)
}